Sources/AsyncHTTPClient/HTTPClient.swift

```diff
@@ -201,31 +217,22 @@ public class HTTPClient {
     ///     - deadline: Point in time by which the request must complete.
     public func execute<Delegate: HTTPClientResponseDelegate>(request: Request,
                                                               delegate: Delegate,
-                                                              eventLoop: EventLoopPreference,
+                                                              eventLoop eventLoopPreference: EventLoopPreference,
                                                               deadline: NIODeadline? = nil) -> Task<Delegate.Response> {
-        switch eventLoop.preference {
+        let taskEL: EventLoop
+        switch eventLoopPreference.preference {
         case .indifferent:
-            return self.execute(request: request, delegate: delegate, eventLoop: self.eventLoopGroup.next(), deadline: deadline)
+            taskEL = self.pool.associatedEventLoop(for: ConnectionPool.Key(request)) ?? self.eventLoopGroup.next()
         case .delegate(on: let eventLoop):
             precondition(self.eventLoopGroup.makeIterator().contains { $0 === eventLoop }, "Provided EventLoop must be part of clients EventLoopGroup.")
-            return self.execute(request: request, delegate: delegate, eventLoop: eventLoop, deadline: deadline)
+            taskEL = eventLoop
         case .delegateAndChannel(on: let eventLoop):
             precondition(self.eventLoopGroup.makeIterator().contains { $0 === eventLoop }, "Provided EventLoop must be part of clients EventLoopGroup.")
-            return self.execute(request: request, delegate: delegate, eventLoop: eventLoop, deadline: deadline)
-        case .testOnly_exact(channelOn: let channelEL, delegateOn: let delegateEL):
-            return self.execute(request: request,
-                                delegate: delegate,
-                                eventLoop: delegateEL,
-                                channelEL: channelEL,
-                                deadline: deadline)
+            taskEL = eventLoop
+        case .testOnly_exact(_, delegateOn: let delegateEL):
```

Comment:
this isn't right. In this case we should just make a fresh connection on `channelEL` with delegate on `delegateEL`

Reply:
Is your comment `testOnly_exact`? If it is I think this is what is happening: by setting `taskEL` to `delegateEL`, the delegate will be called on the `delegateEL`, and the pool takes care of using the correct `EventLoop` for the channel.

Reply:
@adtrevor yes, it’s about testOnly_exact. The point of this preference (which is only for unit tests) is to get back a channel on the selected `channelEL` and a delegate firing on `delegateEL`.
I don’t think it’s possible that the “pool takes care” of putting the channel on `channelEL` because the `channelEL` parameter is actually ignored.

Reply:
@weissi The preference is passed to the provider, it is the responsibility of the `resolvePreference` method:
https://github.com/swift-server/async-http-client/pull/105/files#diff-d4fe528099e126b77e626c998dfbf926R460

To return what would be the best EventLoop for the channel and if it is mandatory or not to use it.

Reply:
@adtrevor ah I see, thank you!